controller/controller.go

```diff
@@ -270,14 +309,44 @@ func (c *Manager) Run(rolloutThreadiness, serviceThreadiness, ingressThreadiness
 		}
 	}
 
-	// Start the informer factories to begin populating the informer caches
-	log.Info("Starting Controllers")
-	go wait.Until(func() { c.rolloutController.Run(rolloutThreadiness, stopCh) }, time.Second, stopCh)
-	go wait.Until(func() { c.serviceController.Run(serviceThreadiness, stopCh) }, time.Second, stopCh)
-	go wait.Until(func() { c.ingressController.Run(ingressThreadiness, stopCh) }, time.Second, stopCh)
-	go wait.Until(func() { c.experimentController.Run(experimentThreadiness, stopCh) }, time.Second, stopCh)
-	go wait.Until(func() { c.analysisController.Run(analysisThreadiness, stopCh) }, time.Second, stopCh)
-	log.Info("Started controller")
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if !electOpts.LeaderElect {
+		log.Info("Leader election is turned off. Running in single-instance mode")
+		go c.startLeading(ctx, rolloutThreadiness, serviceThreadiness, ingressThreadiness, experimentThreadiness, analysisThreadiness)
+	} else {
+		// id used to distinguish between multiple controller manager instances
+		id, err := os.Hostname()
+		if err != nil {
+			log.Fatal("error getting hostname so that the rollouts controllers can't elect a leader")
+		}
+		// add a uniquifier so that two processes on the same host don't accidentally both become active
+		id = id + "_" + string(uuid.NewUUID())
+		log.Infof("leaderelection get id %s", id)
+		go leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+			Lock: &resourcelock.LeaseLock{
+				LeaseMeta: metav1.ObjectMeta{Name: "rollouts-controller", Namespace: electOpts.LeaderElectionNamespace}, Client: c.kubeClientSet.CoordinationV1(),
```

Comment:
To avoid conflict, shall we change the leadMeta.Name to `argo-rollouts-controller-lock` by default and make it configurable like LeaderElectionNamespace?